Add MistServer 3.11

junojense opened the pull request:
# Checklist for Review

**NOTE:** This checklist is intended for the use of the Official Images maintainers both to track the status of your PR and to help inform you and others of where we're at. As such, please leave the "checking" of items to the repository maintainers. If there is a point below for which you would like to provide additional information or note completion, please do so by commenting on the PR. Thanks! (and thanks for staying patient with us :heart:)

-	[x] associated with or contacted upstream?
	- repo under https://github.com/DDVTECH (https://github.com/DDVTECH/mistserver is upstream)
-	[x] available under [an OSI-approved license](https://opensource.org/licenses)?
	- "unlicense" (public domain)
-	[x] does it fit into one of the common categories? ("service", "language stack", "base distribution")
	- service
-	[ ] is it reasonably popular, or does it solve a particular use case well?
-	[x] does a [documentation](https://github.com/docker-library/docs/blob/master/README.md) PR exist? (should be reviewed and merged at roughly the same time so that we don't have an empty image page on the Hub for very long)
	- https://github.com/docker-library/docs/pull/2640
-	[ ] official-images maintainer dockerization review for best practices and cache gotchas/improvements (ala [the official review guidelines](https://github.com/docker-library/official-images/blob/master/README.md#review-guidelines))?
-	[ ] 2+ official-images maintainer dockerization review?
-	[ ] existing official images have been considered as a base? (ie, if `foobar` needs Node.js, has `FROM node:...` instead of grabbing `node` via other means been considered?)
-	[x] ~~if `FROM scratch`, tarballs only exist in a single commit within the associated history?~~
-	[ ] passes current tests? any simple new tests that might be appropriate to add, [assuming this image has any other images it might be interchangeable with](https://github.com/docker-library/official-images/tree/HEAD/test#what-tests-belong-here)?

## Extra Information
I am associated with upstream, the forked repository is contained in the official MistServer organisation. Our software is public domain, and is "licensed" under the [unlicense](https://opensource.org/license/unlicense). The current image on Docker hub builds off alpine and can be found [here](https://hub.docker.com/r/ddvtech/mistserver). The tests seem to be passing. I have also created a [documentation PR](https://github.com/docker-library/docs/pull/2640) which passes the markdown formatter.

junojense committed:
Added mistserver to the library

junojense commented:
Hi again, can this PR be looked at?

tianon commented:
I've taken a first pass at review, and these are my very high-level notes from it:

At first glance, the multi-stage build doesn't seem to be one of the cases in https://github.com/docker-library/faq#multi-stage-builds (using `apk add --virtual` and something like `scanelf` to find the runtime deps and keep them afterwards would probably work really well).

That's a lot of `ARG` - are all these really meaningful? Should their values be documented better? (Our build tooling won't ever use them, so they're mostly just noise here.)

Explicitly using `Builder: buildkit` isn't really necessary (the `TARGETPLATFORM` value it seems to be used to get isn't something we recommend relying on, preferring instead to see userspace/runtime detection based off things like `apk --print-arch`).

Building mbedtls from source is a little eyebrow-raising -- Alpine contains the same version 3.6.5, is there a reason it isn't used? (https://pkgs.alpinelinux.org/package/v3.23/main/x86_64/mbedtls-dev)

Having such _long_ single-line `RUN` lines is really hard to read/review -- those should be split up with `\` and newlines.

Using `FROM alpine` with no specific version is going to have a large tendency to break when the "square wheel" rolls, so we highly recommend pinning to an explicit release of Alpine like `FROM alpine:3.23` so that you can more directly control/manage the rolling of that wheel (and the breakage that may or may not come from it).

Using `LABEL` is not recommended, as the inheritance behavior of it is really poor; see https://github.com/docker-library/official-images/issues/3540, especially https://github.com/docker-library/official-images/issues/3540#issuecomment-2715158488 (our build system automatically adds appropriate similar annotations).

Using `HEALTHCHECK` is also not something I'd personally recommend setting on the image (https://github.com/docker-library/faq#healthcheck) -- perhaps this command should be included in the documentation instead, so that users who want its behavior can get it easily?  (in either a healthcheck in Docker or an appropriate probe in Kubernetes, for example, which won't read `HEALTHCHECK` metadata anyways)

Ideally the download of the source code would have some kind of verification, but that's a little hard with the source code tarball being generated by GitHub directly. 😔

junojense committed:
Removed Builder:buildkit and updated to MistServer 3.11.1

junojense commented:
Thanks for the feedback @tianon, all points have been addressed in the updated Dockerfile.

- Got rid of the multi-stage build, we now use `scanelf` to find run-time deps
- Significantly reduced `ARG` usage, the remaining arguments are...
    - `MIST_DEBUG` sets the MistServer logging level to production instead of development
    - `MIST_OPTS` space separated compile options as specified [here](https://docs.mistserver.org/mistserver/installation/compile#compile-flags)
- Removed `Builder: buildkit` in favour of `apk --print-arch`
- Removed `HEALTHCHECK`, `LABEL`, improved the formatting of `RUN`, and pinned the alpine version

To elaborate on building mbedtls from source; we require DTLS support which is not present in mbedtls official builds / releases. Though, the retrieved source code is now at least verified with its SHA-256 checksum.

As for verification of MistServer itself, our newest release (3.11.1) includes [downloads](https://mistserver.org/download) for the source code with matching SHA-256 checksum that we check against in the new Dockerfile. 

As everything mentioned has been addressed, the PR should be ready for further review.
